### ISSUE TITLE
Add product nodes to DSMGP

### ISSUE THREAD

**Issue by trappmartin** [stevie1023/cspn-gpspn#1]
Hi,
I would suggest extending the code for the DSMGP to contain product nodes as defined in case of a conditional SPN.
Currently, you seem to have sum nodes: 
https://github.com/stevie1023/cspn-gpspn/blob/f3416ff840c2883885a69476530b1442deaa119d/cspn22/cspn2-master/originalspngp.py#L16-L80

and product nodes that split the covariate space:
https://github.com/stevie1023/cspn-gpspn/blob/f3416ff840c2883885a69476530b1442deaa119d/cspn22/cspn2-master/originalspngp.py#L83-L176

defined in your model.

In addition, we would need an implementation of product nodes that split the output space:
https://github.com/stevie1023/cspn-gpspn/blob/f3416ff840c2883885a69476530b1442deaa119d/cspn22/cspn2-master/new_base.py#L130-L141

If you don't know how to do this, you might want to as Alejandro Molina or Zhongjie.

Once you have added the additional type, you should be able to adjust the structure construction. The current code seems to use a region-graph on top of the model, this is not necessary. You can simplify the implementation by constructing the SPN (DSMGP) directly.

Here is how this is implemented in my Julia codebase:
https://github.com/trappmartin/DeepStructuredMixtures/blob/bf43a03ee5f78ed8dda903cd4dd88651531dde4b/src/treeStructure.jl#L131-L307

It is essentially a recursive algorithm. Here is an illustrative example of how this could be implemented (pseudo-code). I'm not a python person, so this might have to be implemented slightly different but it should give you an idea of how it works.

```python
def build_structure(X, Y):  # X and Y are the observations
  XSpace = # determine covariate space from X, e.g. [-inf, inf]
  YSpace = # determine countable output space from Y, e.g. [Y1, Y2, Y3]
  return build_sum_node(X, Y, XSpace, YSpace)

def build_sum_node(X, Y, XSpace, YSpace):
  children = [build_split_node(X, Y, XSpace, YSpace) for k in 1:K] # build children
  w = [1/K for k in 1:K]
  return Sum(children, w)

def get_covariate_partition(XSpace):
  return # return a partitioning of the input space (see getsplits)

def build_split_node(X, Y, XSpace, YSpace):
  children = []
  partition = get_covariate_partition(XSpace)
  for part in partition:
    Xpart = # select covariates of observations in part
    Ypart = # select ouputs of observations in part

    if number of observations in part is smaller than M:
      append(children, build_leaf(Xpart, Ypart))
    else:
      append(children, build_product_node(Xpart, Ypart, part, YSpace)
  return Split(children, partition)

def get_output_partition(XSpace):
  return # return a partitioning of the output space, e.g. random

def build_product_node(X, Y, XSpace, YSpace):
  children = []
  partition = get_output_partition(YSpace)
  children = [build_sum_node(X, Y, XSpace, part) for part in partition]
  return Product(children, partition)

def build_leaf(X, Y):
  if Y is multivariate:
    leaf = Product([GP(X, yi) for yi in Y], ...) ... fully factorise Y
  else:
    leaf = GP(X, Y)
  return leaf
```

**Comment by stevie1023** [comment]
Hi Martin,
Thanks for your detailed explanation. Now I get more sense out of it. So you would suggest me to write a new .py file to combine all the functions you defined in the pseudo-code and include corresponding nodes? Seems quite straightforward:)
Thanks again~

Best,
stevie

发件人: Martin Trapp<mailto:notifications@github.com>
发送时间: 2020年9月21日 19:25
收件人: stevie1023/cspn-gpspn<mailto:cspn-gpspn@noreply.github.com>
抄送: Subscribed<mailto:subscribed@noreply.github.com>
主题: [stevie1023/cspn-gpspn] Add product nodes to DSMGP (#1)


Hi,
I would suggest extending the code for the DSMGP to contain product nodes as defined in case of a conditional SPN.
Currently, you seem to have sum nodes:
https://github.com/stevie1023/cspn-gpspn/blob/f3416ff840c2883885a69476530b1442deaa119d/cspn22/cspn2-master/originalspngp.py#L16-L80

and product nodes that split the covariate space:
https://github.com/stevie1023/cspn-gpspn/blob/f3416ff840c2883885a69476530b1442deaa119d/cspn22/cspn2-master/originalspngp.py#L83-L176

defined in your model.

In addition, we would need an implementation of product nodes that split the output space:
https://github.com/stevie1023/cspn-gpspn/blob/f3416ff840c2883885a69476530b1442deaa119d/cspn22/cspn2-master/new_base.py#L130-L141

If you don't know how to do this, you might want to as Alejandro Molina or Zhongjie.

Once you have added the additional type, you should be able to adjust the structure construction. The current code seems to use a region-graph on top of the model, this is not necessary. You can simplify the implementation by constructing the SPN (DSMGP) directly.

Here is how this is implemented in my Julia codebase:
https://github.com/trappmartin/DeepStructuredMixtures/blob/bf43a03ee5f78ed8dda903cd4dd88651531dde4b/src/treeStructure.jl#L131-L307

It is essentially a recursive algorithm. Here is an illustrative example of how this could be implemented (pseudo-code). I'm not a python person, so this might have to be implemented slightly different but it should give you an idea of how it works.

def build_structure(X, Y):  # X and Y are the observations

  XSpace = # determine covariate space from X, e.g. [-inf, inf]

  YSpace = # determine countable output space from Y, e.g. [Y1, Y2, Y3]

  return build_sum_node(X, Y, XSpace, YSpace)



def build_sum_node(X, Y, XSpace, YSpace):

  children = [build_split_node(X, Y, XSpace, YSpace) for k in 1:K] # build children

  w = [1/K for k in 1:K]

  return Sum(children, w)



def get_covariate_partition(XSpace):

  return # return a partitioning of the input space (see getsplits)



def build_split_node(X, Y, XSpace, YSpace):

  children = []

  partition = get_covariate_partition(XSpace)

  for part in partition:

    Xpart = # select covariates of observations in part

    Ypart = # select ouputs of observations in part



    if number of observations in part is smaller than M:

      append(children, build_leaf(Xpart, Ypart))

    else:

      append(children, build_product_node(Xpart, Ypart, part, YSpace)

  return Split(children, partition)



def get_output_partition(XSpace):

  return # return a partitioning of the output space, e.g. random



def build_product_node(X, Y, XSpace, YSpace):

  children = []

  partition = get_output_partition(YSpace)

  children = [build_sum_node(X, Y, XSpace, part) for part in partition]

  return Product(children, partition)



def build_leaf(X, Y):

  if Y is multivariate:

    leaf = Product([GP(X, yi) for yi in Y], ...) ... fully factorise Y

  else:

    leaf = GP(X, Y)

  return leaf

―

**Comment by trappmartin** [comment]
Whatever is easier, but it feels that implementing a new clean version of the structure construction would be beneficial for you and the codebase. :)

**Comment by stevie1023** [comment]
Hi Martin,
After discussion with Zhongjie, I want to give my understanding more specifically in case I get something wrong again.
So sum nodes have split nodes as children and split nodes have their own XSpace (on which dimension/channel they split), so XSpace is essentially Xi, or part of Xi after splitted and keeps changing ? Am I right here?
If so, [cid:image001.png@01D69120.823AF050]  I am not sure what should  the partition look like? They have the same D as input X, but less samples(smaller N) or only one dimension?[cid:image004.png@01D69124.20F42650]
Thanks/😊
发件人: Martin Trapp<mailto:notifications@github.com>
发送时间: 2020年9月22日 17:20
收件人: stevie1023/cspn-gpspn<mailto:cspn-gpspn@noreply.github.com>
抄送: Stevie Zhu<mailto:zzmmyy1023@outlook.com>; Comment<mailto:comment@noreply.github.com>
主题: Re: [stevie1023/cspn-gpspn] Add product nodes to DSMGP (#1)


Whatever is easier, but it feels that implementing a new clean version of the structure construction would be beneficial for you and the codebase. :)

**Comment by trappmartin** [comment]
I cannot see the image you refer to. Could you please change your comment directly on GitHub so that it is visible? Thanks.

**Comment by stevie1023** [comment]
oh sorry.
![image](https://user-images.githubusercontent.com/51037813/93891999-9f333180-fd1e-11ea-84bd-d43fd7e1cc60.png)

this is the image

**Comment by stevie1023** [comment]
![1](https://user-images.githubusercontent.com/51037813/93892567-3e582900-fd1f-11ea-8e23-1634a6da02a5.png)
Sorry, this is the iamge. I hope it is visible now.

**Comment by trappmartin** [comment]
Sorry for the late reply.

I think you are on the right track, but I'm not 100% sure about it so I decided to make a small illustration myself. :)

![grafik](https://user-images.githubusercontent.com/7974003/94339852-73090080-fffd-11ea-8322-2d17ccb7c20d.png)

So that you fully understand the example I'll try to explain it a little bit.

What we want at the end of the day is to learn a function/map from a space \mathcal{X} to a space \mathcal{Y}. In the pseudo-code I denoted those as XSpace and YSpace.
In order to learn that function, we will use data that we observed, namely bold X and bold Y. 

For simplicity, let's ignore the data for now as it is not relevant in order to understand the SPN structure. The structure is in fact somewhat data independent.

What I propose is to use an SPN that recursively partitions the input space XSpace and the output space YSpace. You see in the illustration that the SPN has a sum node as root node. This node represents a function from XSpace to YSpace (or \mathcal{X} to \mathcal{Y}). Now we construct split nodes (which are essentially product nodes) under this sum node. In the example, I assume that XSpace is a one-dimensional real line, but it can be higher-dimensional too. It doesn't matter. The key is that we partition XSpace/\mathcal{X} into disjoint parts. If we partition XSpace into two parts, we obtain two functions one mapping from XSpace_1 to YSpace and another from XSpace_2 to YSpace. The union of those two input spaces, e.g. XSpace_1 \cup XSpace_2, is again XSpace. 
After this partition, we construct product nodes that split the output space (YSpace). As YSpace is generally a finite countable set, this is akin to what you know from SPN literature. Then afterwards we construct as sum node and so on...

The data bold X and bold Y are *implicitly* split into parts. We don't actually partition bold X and bold Y, but this happens implicitly through the partitioning of XSpace and YSpace. In the end, each leaf in the network (which is a Gaussian process), will be conditioned on a sub-set of bold X and bold Y and will only be a function from some sub-space XSpace_i to a sub-space YSpace_i. And we can always build the SPN such that YSpace_i will only be a single output dimension.

**Comment by trappmartin** [comment]
Let me know if something is not clear. We can also have a Skype call about it, which might be easier to answer your immediate questions.

**Comment by minimrbanana** [comment]
Hi Martin, as far as I understand, we will have recursively #sum (mixture of its children) --> #split-X (split the X space) -->#product-Y (split the Y space) -->sum .....
in the leaf node we model the GP of Y_i given subspace of X.
Is this identical to what you proposed? Just want to make sure I understand you correctly.

**Comment by trappmartin** [comment]
Yes, that’s exactly what I was trying to say. :)

**Comment by trappmartin** [comment]
What's the status of this?

**Comment by minimrbanana** [comment]
Hi Martin, Stevie was busy with applying for her Masters study, I think it is almost done and she should be back very soon

**Comment by trappmartin** [comment]
Perfect.

Good luck with the application Steve!

**Comment by stevie1023** [comment]
Hi guys I'm finally back to work... Sorry for keeping you waiting these days. Now I'm done with my application(amen= =). I start to work right away~

**Comment by stevie1023** [comment]
<img width="216" alt="Snipaste_2020-10-07_18-48-12" src="https://user-images.githubusercontent.com/51037813/95321506-b43ac380-08cd-11eb-8c84-799612a175d6.png">

Hi，so sorry for the delay(I was on vocation lol). This is how the structure currently looks like. Maybe you can have a look and make sure it makes sense:)
Here is my logic: I set a threshold (min_idx) for the product node, if the idx < min_idx, the product node creates several gp leaves, each accounting for one channel of y(split y); if idx >= min_idx and y.shape[1]>1, the product node creates two sum as children, whose children is split node. The process goes on recursively until the final idx meets the criterion. Y is split by product nodes and each gp leave models one y channel.

Now I am working on passing the corresponding y observations to each gp leave for training. But I want to make sure the sturcture above makes sense or not:)

cheers~

**Comment by minimrbanana** [comment]
It looks almost correct in my opinion. One question, also to @trappmartin , in the 4th layer (+) in the above figure, do we suppose to have multiple children(like the root +) or one child is already enough?

**Comment by trappmartin** [comment]
Yes, this looks almost perfect. Each sum node should have the same number of children. So also the sum node in the 4th layer should have multiple children. In general it’s best to have at least two children under each sum node.

**Comment by trappmartin** [comment]
Do split nodes have always only two children? In my experience it is useful to have more children in order to scale down the computational costs.

**Comment by stevie1023** [comment]
Hi, currently my structure is modified to following, each sum node has two children(split), each split node has two product node as children(we can have more children later). Each gp leave is used to model one scope of y. Now I am working on the computation part to come up with the final result. Wish me luck!
[cid:image001.png@01D69DA4.AE2201F0]

发件人: Martin Trapp<mailto:notifications@github.com>
发送时间: 2020年10月8日 4:44
收件人: stevie1023/cspn-gpspn<mailto:cspn-gpspn@noreply.github.com>
抄送: Stevie Zhu<mailto:zzmmyy1023@outlook.com>; Comment<mailto:comment@noreply.github.com>
主题: Re: [stevie1023/cspn-gpspn] Add product nodes to DSMGP (#1)


Do split nodes have always only two children? In my experience it is useful to have more children in order to scale down the computational costs.

―

**Comment by trappmartin** [comment]
Sounds good.

Let me know if you need to help in the computational part.
Here are some points that should you get started:

Computation at sum nodes: https://github.com/trappmartin/DeepStructuredMixtures/blob/bf43a03ee5f78ed8dda903cd4dd88651531dde4b/src/common.jl#L275-L292

Computation at split nodes: https://github.com/trappmartin/DeepStructuredMixtures/blob/bf43a03ee5f78ed8dda903cd4dd88651531dde4b/src/common.jl#L243-L254

Computation at product nodes: https://math.stackexchange.com/questions/147497/covariance-matrix-of-mixture-distribution

Let me know if you have questions.

**Comment by trappmartin** [comment]
I just realise I messed up the pointers. Sorry.

For one observation you do as follows:
1. Split nodes act as gates, meaning they return the mean and covariance of exactly one of its children. The child is selected based on the partition and the location of the observation.
2. Product nodes concatenate the mean vector and the diagonal or block-diagonal of the covariance matrix. 
2. Sum nodes compute a weighted average for the mean vector and compute their covariance matrix according to the stack exchange link I posted above.

**Comment by stevie1023** [comment]
Oh nice. That's my original understanding of the whole computational part.
the whole inference is almost done and only a few codes need to be modified. Hope I can get it done soon!

**Comment by stevie1023** [comment]
Hi, I think I figured it out! So the rough structure and computation are done  which seem to work out nicely. The result seems nice( no solid check yet:). What is our next move? Run it on several datasets and compare with other measures?

**Comment by trappmartin** [comment]
Yes, next move is to evaluate the NLPD on various datasets and compare it to a structure that doesn’t have the product nodes but simply factorises Y at the root. Your structure should perform better.

**Comment by minimrbanana** [comment]
Hi Martin, Stevie has implemented the conditional independence test based on the CSPN code, in order to compare with Random-split of Y. We did a sanity check compared with this package(https://cran.r-project.org/web/packages/CondIndTests/CondIndTests.pdf) and it seems the implementation works. However, in practice, when operated on our datasets --windmill and Parkin.-- it turns out the Ys are not enough independent given X. That is, given X, Y_i and Y_j are highly correlated, resulting in product nodes happening (almost) only at leaves. 
On the other hand, if we force Y to split randomly, in the case of Y in 2d, it will just split at the very top of our Tree, do you see if this can be an issue?
Thus I think we need to figure out a strategy of product node splitting. (Force random split works anyway)

**Comment by trappmartin** [comment]
I honestly don’t have a good intuition for that. Splits further down the tree might be better but I honestly don’t know. I think this is something we should investigate to find a good strategy.

Is the CSPN independence Test a G Test?

**Comment by minimrbanana** [comment]
The implemented method we use is called RCoT from (https://arxiv.org/pdf/1702.03877.pdf)

**Comment by trappmartin** [comment]
Thanks, I’ll have a look at it next week. I’m currently moving.
